`-Wtemplate-id-cdtor` compilation warnings with GCC 14

### Tested versions

- Reproducible in 4.3.dev (86bf8354a06ab7b23a0ff6a81b48fd015e92ac94)
- Likely affects earlier branches too so the fix should be cherry-picked

### System information

Fedora Linux 40 (KDE Plasma) - Wayland - Vulkan (Forward+) - dedicated AMD Radeon RX 7600M XT (RADV NAVI33) () - AMD Ryzen 7 7840HS w/ Radeon 780M Graphics (16 Threads)

### Issue description

I just upgraded to Fedora 40 which provides GCC 14.0.1 (RC version).

This version started raising a new type of warning:
```
warning: template-id not allowed for constructor in C++20 [-Wtemplate-id-cdtor]
```
We don't compile with C++20 yet (@Repiteo is working on supporting it in #89660), but it seems like GCC made this a warning for earlier versions of the standard anyway.

Here are the affected files from a Linux editor dev build:
```
./core/templates/safe_refcount.h:149:49: warning: template-id not allowed for constructor in C++20 [-Wtemplate-id-cdtor]
  149 |         _ALWAYS_INLINE_ explicit SafeNumeric<T>(T p_value = static_cast<T>(0)) {
      |                                                 ^
./core/templates/safe_refcount.h:149:49: note: remove the '< >'
```
```
./core/templates/ring_buffer.h:214:22: warning: template-id not allowed for constructor in C++20 [-Wtemplate-id-cdtor]
  214 |         RingBuffer<T>(int p_power = 0) {
      |                      ^
./core/templates/ring_buffer.h:214:22: note: remove the '< >'
./core/templates/ring_buffer.h:217:9: warning: template-id not allowed for destructor in C++20 [-Wtemplate-id-cdtor]
  217 |         ~RingBuffer<T>() {}
      |         ^
./core/templates/ring_buffer.h:217:9: note: remove the '< >'
```
```
thirdparty/clipper2/include/clipper2/clipper.core.h:183:22: warning: template-id not allowed for constructor in C++20 [-Wtemplate-id-cdtor]
  183 |     explicit Point<T>(const Point<T2>& p) { Init(p.x, p.y); }
      |                      ^
thirdparty/clipper2/include/clipper2/clipper.core.h:183:22: note: remove the '< >'
```

Workaround for affected users:
- Either build with `werror=no`
- Or add `cxxflags="-Wno-error=template-id-cdtor"` to your scons command

### Steps to reproduce

- `scons dev_build=yes dev_mode=yes`

### Minimal reproduction project (MRP)

n/a

--- a/core/templates/ring_buffer.h
+++ b/core/templates/ring_buffer.h
@@ -211,10 +211,10 @@ public:
 		size_mask = mask;
 	}
 
-	RingBuffer<T>(int p_power = 0) {
+	RingBuffer(int p_power = 0) {
 		resize(p_power);
 	}
-	~RingBuffer<T>() {}
+	~RingBuffer() {}
 };
 
 #endif // RING_BUFFER_H

--- a/core/templates/safe_refcount.h
+++ b/core/templates/safe_refcount.h
@@ -146,7 +146,7 @@ public:
 		}
 	}
 
-	_ALWAYS_INLINE_ explicit SafeNumeric<T>(T p_value = static_cast<T>(0)) {
+	_ALWAYS_INLINE_ explicit SafeNumeric(T p_value = static_cast<T>(0)) {
 		set(p_value);
 	}
 };

--- a/thirdparty/clipper2/include/clipper2/clipper.core.h
+++ b/thirdparty/clipper2/include/clipper2/clipper.core.h
@@ -138,7 +138,7 @@ namespace Clipper2Lib
     }
 
     template <typename T2>
-    explicit Point<T>(const Point<T2>& p)
+    explicit Point(const Point<T2>& p)
     {
       Init(p.x, p.y, p.z);
     }
@@ -180,7 +180,7 @@ namespace Clipper2Lib
     Point(const T2 x_, const T2 y_) { Init(x_, y_); }
 
     template <typename T2>
-    explicit Point<T>(const Point<T2>& p) { Init(p.x, p.y); }
+    explicit Point(const Point<T2>& p) { Init(p.x, p.y); }
 
     Point operator * (const double scale) const
     {

--- a/thirdparty/clipper2/patches/gcc14-warning.patch
+++ b/thirdparty/clipper2/patches/gcc14-warning.patch
@@ -1,0 +1,22 @@
+diff --git a/thirdparty/clipper2/include/clipper2/clipper.core.h b/thirdparty/clipper2/include/clipper2/clipper.core.h
+index a77cdad5f4..0de7c3720e 100644
+--- a/thirdparty/clipper2/include/clipper2/clipper.core.h
++++ b/thirdparty/clipper2/include/clipper2/clipper.core.h
+@@ -138,7 +138,7 @@ namespace Clipper2Lib
+     }
+ 
+     template <typename T2>
+-    explicit Point<T>(const Point<T2>& p)
++    explicit Point(const Point<T2>& p)
+     {
+       Init(p.x, p.y, p.z);
+     }
+@@ -180,7 +180,7 @@ namespace Clipper2Lib
+     Point(const T2 x_, const T2 y_) { Init(x_, y_); }
+ 
+     template <typename T2>
+-    explicit Point<T>(const Point<T2>& p) { Init(p.x, p.y); }
++    explicit Point(const Point<T2>& p) { Init(p.x, p.y); }
+ 
+     Point operator * (const double scale) const
+     {

